lib/Dialect/Arc/ArcCostModel.cpp

```diff
@@ -0,0 +1,120 @@
+//===- ArcCostModel.cpp ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Arc/ArcCostModel.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include <algorithm>
+
+using namespace llvm;
+using namespace circt;
+using namespace arc;
+using namespace std;
+
+size_t ArcCostModel::getCost(Operation *op) { return computeOperationCost(op); }
+
+size_t ArcCostModel::computeOperationCost(Operation *op) {
+  if (opCostCash.count(op))
+    return opCostCash[op];
+  if (isa<circt::comb::ConcatOp>(op))
+    return opCostCash[op] = size_t(OperationCost::CONCATCOST);
+  if (isa<circt::comb::ExtractOp>(op))
+    return opCostCash[op] = size_t(OperationCost::EXTRACTCOST);
+  // We have some other functions that need to be handled in a different way
+  // arc::StateOp, arc::CallOp, mlir::func::CallOp and arc::VectorizeOp, each of
+  // these functions have bodies so the cost of the op equals the cost of its
+  // body.
+  if (isa<arc::StateOp>(op) || isa<arc::CallOp>(op) ||
+      isa<mlir::func::CallOp>(op)) {
+    size_t totalCost = 0;
+    const auto regions =
+        dyn_cast<CallOpInterface>(op).resolveCallable()->getRegions();
+    for (auto &region : regions)
+      for (auto &block : region)
+        for (auto &innerOp : block)
+          totalCost += computeOperationCost(&innerOp);
+    return opCostCash[op] = totalCost;
+  }
```

Comment:
In this case I would do something like `getCost(cast<CallOpInterface>(op).resolveCallable())` and then make the `for region { for block { for innerOp { count cost }}}` pattern the default for any operation. That will allow you to call `getCost()` on an `hw::HWModuleOp` or `mlir::ModuleOp` or `arc::DefineOp` as well, and get the correct results.

Reply:
I think I don't get this. should I call `getCost` as you suggest?
but this will make a recursive chain as `getCost` will call `computeOperationCost`?
so the `computeOperationCost` will just have the base cases and the `nested for loops`?
should we remove the check for `vectorizeOp` as well?

Reply:
Oh you can also call `computeOperationCost` instead, since `getCost` is just forwarding the call. But yeah, I would call `computeOperationCost` on the thing being called, and then add a way to compute the cost of function and arc definitions. That will make the cost model easier to use, because you can just ask for the cost of an entire function or arc.

I think you still need a bit of special handling for the vectorize op, since it has the shuffling cost you need to account for.

Reply:
> I would call computeOperationCost on the thing being called, and then add a way to compute the cost of function and arc definitions

Please explain this part, it's a bit confusing to me.

Reply:
How about this?
```
size_t ArcCostModel::computeOperationCost(Operation *op) {
  if (opCostCash.count(op))
    return opCostCash[op];
  if (isa<circt::comb::ConcatOp>(op))
    return opCostCash[op] = size_t(OperationCost::CONCATCOST);
  if (isa<circt::comb::ExtractOp>(op))
    return opCostCash[op] = size_t(OperationCost::EXTRACTCOST);

  if (isa<arc::VectorizeOp>(op)) {
    size_t inputVecCost = getInputVectorsCost(dyn_cast<VectorizeOp>(op));
    size_t vecOpBodyCost = 0;
    auto regions = op->getRegions();
    for (auto &region : regions)
      for (auto &block : region)
        for (auto &innerOp : block)
          vecOpBodyCost += computeOperationCost(&innerOp);

    vectoroizeOpsBodyCost += vecOpBodyCost;
    allVectorizeOpsCost += inputVecCost + vecOpBodyCost;
    return opCostCash[op] = inputVecCost + vecOpBodyCost;
  }
  // We have some other functions that need to be handled in a different way
  // arc::StateOp, arc::CallOp, mlir::func::CallOp and arc::VectorizeOp, each of
  // these functions have bodies so the cost of the op equals the cost of its
  // body.
  if (auto *callableOp = cast<CallOpInterface>(op).resolveCallable()) {
    size_t totalCost = 0;
    const auto regions = callableOp->getRegions();
    for (auto &region : regions)
      for (auto &block : region)
        for (auto &innerOp : block)
          totalCost += computeOperationCost(&innerOp);
    return opCostCash[op] = totalCost;
  }

  return opCostCash[op] = size_t(OperationCost::NORMALCOST);
}
```

Reply:
I would turn this:
```c++
  if (auto *callableOp = cast<CallOpInterface>(op).resolveCallable()) {
    size_t totalCost = 0;
    const auto regions = callableOp->getRegions();
    for (auto &region : regions)
      for (auto &block : region)
        for (auto &innerOp : block)
          totalCost += computeOperationCost(&innerOp);
    return opCostCash[op] = totalCost;
  }
```
into something like this:
```c++
  if (auto callableOp = dyn_cast<CallOpInterface>(op))
    if (auto *calledOp = callableOp.resolveCallable())
      return computeOperationCost(*calledOp);
  if (isa<func::FuncOp, arc::DefineOp, mlir::ModuleOp>(op)) {
    size_t totalCost = 0;
    for (auto &region : op.getRegions())
      for (auto &block : region)
        for (auto &innerOp : block)
          totalCost += computeOperationCost(&innerOp);
    return opCostCash[op] = totalCost;
  }
```
And then switch from `size_t` costs to the `OperationCosts` struct you've proposed above 😃. This makes call ops use `computeOperationCost` on the op they are calling. And the region->block->ops cost counting is done whenever you call `computeOperationCost` on a `func.func`, `arc.define`, or `builtin.module`. This will allow you to compute the total operation cost in your IR, e.g. for printing, by just asking your cost model for the cost of the root `builtin.module` op.